### PR TITLE
allow arbitrary inherent impls for builtin types in core

### PR DESCRIPTION
**Pull request by lcnr**
Part of https://github.com/rust-lang/compiler-team/issues/487. Slightly adjusted after some talks with @m-ou-se about the requirements of `t-libs-api`.

This adds a crate attribute `#![rustc_coherence_is_core]` which allows arbitrary impls for builtin types in core.

For other library crates impls for builtin types should be avoided if possible. We do have to allow the existing stable impls however. To prevent us from accidentally adding more of these in the future, there is a second attribute `#[rustc_allow_incoherent_impl]` which has to be added to **all impl items**. This only supports impls for builtin types but can easily be extended to additional types in a future PR.

This implementation does not check for overlaps in these impls. Perfectly checking that requires us to check the coherence of these incoherent impls in every crate, as two distinct dependencies may add overlapping methods. It should be easy enough to detect if it goes wrong and the attribute is only intended for use inside of std.

The first two commits are mostly unrelated cleanups.

### CODE REVIEW

**Comment by oli-obk** [comment]
the changes lgtm (including query & metadata changes). Try running perfbot once rustdoc builds, too

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 1]
```diff
@@ -1813,6 +1818,33 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
         self.lazy(&all_impls)
     }
 
+    fn encode_incoherent_impls(&mut self) -> Lazy<[IncoherentImpls]> {
+        debug!("EncodeContext::encode_traits_and_impls()");
```
Comment (fbstj) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1822]:
```suggestion
        debug!("EncodeContext::encode_incoherent_impls()");
```
?

**Comment by petrochenkov** [comment]
Any specific reason to put this attribute on the whole crate rather than on specific `impl` items?

The duplication between `rustc_coherence_is_core` and `rustc_allow_incoherent_impl` doesn't look necessary to me.
(And placing `rustc_allow_incoherent_impl` on all functions inside impls instead of the impls themselves looks even more unnecessary.)

If `#[rustc_incoherent_impl]` were a feature-gated attribute applicable to individual impls, then `#![feature(rustc_incoherent_impl)]` would serve the same purpose as `#![rustc_coherence_is_core]`.

**Comment by lcnr** [comment]
After talking with @m-ou-se about this, my takeaways were:

- libs wants to easily add impls for builtin types in core
- ideally `std` is as close to an ordinary library as possible, so while core has a lot of lang items and what not, ideally std does not have any of that and instead only has platform specific + general purpose stuff
- currently there are some impls for builtin types in std, quite a few were added by mistakes and should ideally not exist. t-libs-api wants to be really careful with adding any such impls in the future

So the reasons behind my approach are:

- we have different ways to add incoherent impls because libs has different requirements for these impls.
- `#![rustc_coherence_is_core]` is added so that builtin types are as nice to use as possible in core. Having the attribute on impls would add a mental overhead which shouldn't be needed as we should treat builtin types as being owned by core. Impl wise supporting both options is equally challenging
- `#[rustc_incoherent_impl]` is placed on each item in the impl to make it easier to detect accidental stabilizations of new "incorrect" methods in std or alloc. As adding a method to an existing impl block now also requires this special attribute, it should be a lot more visible during review that something *special* is going on here.
- i use `#![rustc_coherence_is_core]` instead of `#![feature(inherent_impls_for_builtins)]` because 1) features are slightly more work to add 2) features tend to seem like they might be "usable" by ordinary users, which is not the intent here. Using `rustc_attrs` makes it clear that this is purely for internal use.

**Comment by lcnr** [comment]
@petrochenkov this now moves #93766 from adding lang items to adding all the incoherent impls instead. Considered keeping it for lang items as well in case we ever add another lang item applied to an - otherwise untracked - impl but that felt pretty unlikely so I removed the support for that. Does that seem right to you?

**Comment by oli-obk** [comment]
I find it interesting that this is not just a (trivial and ignorable) regression due to the new query, but a much larger one (still tiny at 0.3%) in typeck.

**Comment by lcnr** [comment]
I kind of expect that this happens because there are now a lot more attributes, at least that's why I expect. Might be interesting to somehow filter irrelevant attributes before encoding them in the crate metadata. Idk how much that changes

**Comment by lcnr** [comment]
so pretty much all of the negative perf impact is caused by the additional attributes. Going to look into

> Might be interesting to somehow filter irrelevant attributes before encoding them in the crate metadata

**Comment by oli-obk** [comment]
Maybe add that filter in a separate PR?

I'd be fine merging this before your test commit and working on the attribute thing separately

**Review by oli-obk — COMMENTED** [review]
Modulo the clippy FIXME, this lgtm. What are the plans for that ?

**Comment by lcnr** [comment]
> Modulo the clippy FIXME, this lgtm. What are the plans for that ?

I would guess:

match on `name` to get all supported primitives.
change the return type to `impl Iterator<Item = DefId>` and look over all impls.

**Comment by lcnr** [comment]
@oli-obk implemented my idea

**Review by oli-obk — APPROVED** [review]

**Review by m-ou-se — APPROVED** [review]
The library changes in this PR look good. :) Thanks!

**Comment by lcnr** [comment]
@bors r+ rollup=never

the small perf impact is expected and should imo be fixed separately by changing the way we handle attributes. I intend to take a look at that over the next few weeks.

**Comment by eddyb** [comment]
> crate attribute `#![rustc_coherence_is_core]`

Why not "simply" make `core` (the whole crate) a lang item? I feel like most of this PR could've been avoided. Lang items already are robustly checked, why add more machinery?

File: compiler/rustc_typeck/src/check/method/suggest.rs [hunk 2]
```diff
@@ -281,25 +281,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        || found_assoc(tcx.types.i32)
+                        || found_assoc(tcx.types.i64)
+                        || found_assoc(tcx.types.i128)
+                        || found_assoc(tcx.types.u8)
+                        || found_assoc(tcx.types.u16)
+                        || found_assoc(tcx.types.u32)
+                        || found_assoc(tcx.types.u64)
+                        || found_assoc(tcx.types.u128)
+                        || found_assoc(tcx.types.f32)
+                        || found_assoc(tcx.types.f32);
```
Comment (sunmy2019) [inline compiler/rustc_typeck/src/check/method/suggest.rs:305]: This line looks like a typo to me, should it be `f64`?

Comment (sunmy2019) [reply compiler/rustc_typeck/src/check/method/suggest.rs:305]: Tracked by #114683

**Pull request merged**